src/Type/Doctrine/Descriptors/Carbon/CarbonImmutableType.php

```diff
@@ -0,0 +1,28 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Doctrine\Descriptors\Carbon;
+
+use Carbon\CarbonImmutable;
+use PHPStan\Type\Doctrine\Descriptors\DoctrineTypeDescriptor;
+use PHPStan\Type\ObjectType;
+use PHPStan\Type\Type;
+
+class CarbonImmutableType implements DoctrineTypeDescriptor
+{
+
+	public function getType(): string
+	{
+		return \Carbon\Doctrine\CarbonImmutableType::class;
+	}
+
+	public function getWritableToPropertyType(): Type
+	{
+		return new ObjectType(CarbonImmutable::class);
+	}
+
+	public function getWritableToDatabaseType(): Type
+	{
+		return new ObjectType(CarbonImmutable::class);
```

Comment:
I'm missing this logic here:

> The return type of convertToPHPValue will be the class instance returned by getCarbonClassName, which gets overridden in here.

Reply:
you're right. Is this why the tests are failing or did I miss something registering the descriptor?

Reply:
> The return type of convertToPHPValue will be the class instance returned by getCarbonClassName, which gets overridden in here.

Regarding this, I'm not sure it's needed here, everything seems to work as intended now in the tests (added valid case too).

However after playing around a bit with the code in the library itself I reckon I could make it work using generics: https://phpstan.org/r/2a5b7ba7-681b-4a93-b6e7-19ae2fba7e3c. Is the `ReflectionDescriptor` able to figure out the type through that?

Reply:
Try it :)

Reply:
update on this: it works 🎉 . I will try to get a PR on nesbot/carbon with those annotations, keeping this open until then. 

After that, can this PR just add these?
```
services:
    -
        factory: PHPStan\Type\Doctrine\Descriptors\ReflectionDescriptor('Carbon\Doctrine\CarbonImmutableType')
        tags: [phpstan.doctrine.typeDescriptor]
    -
        factory: PHPStan\Type\Doctrine\Descriptors\ReflectionDescriptor('Carbon\Doctrine\CarbonType')
        tags: [phpstan.doctrine.typeDescriptor]

Reply:
[the PR in Carbon](https://github.com/briannesbitt/Carbon/pull/2344) was merged but I'll wait for the release until re-submitting the ReflectionDescriptor additions